libr/anal/type.c

```diff
@@ -384,3 +385,302 @@ R_API RAnalBaseType *r_anal_get_base_type(RAnal *anal, const char *name) {
+	char *kind = "struct";
+
+	char *sname = r_str_sanitize_sdb_key (type->name);
+	// name=struct
+	sdb_set (anal->sdb_types, sname, kind, 0);
+
+	RStrBuf arglist;
+	RStrBuf param_key;
+	RStrBuf param_val;
+	r_strbuf_init (&arglist);
+	r_strbuf_init (&param_key);
+	r_strbuf_init (&param_val);
+
+	int i = 0;
+	RAnalStructMember *member;
+	r_vector_foreach (&type->struct_data.members, member) {
+		// struct.name.arg1=type,offset,???
+		char *member_sname = r_str_sanitize_sdb_key (member->name);
+		r_strbuf_setf (&param_key, "%s.%s.%s", kind, sname, member_sname);
+		r_strbuf_setf (&param_val, "%s,%" PFMT64u ",%" PFMT64u "", member->type, member->offset, 0);
+		sdb_set (anal->sdb_types, r_strbuf_get (&param_key), r_strbuf_get (&param_val), 0);
+		free (member_sname);
+
+		r_strbuf_appendf (&arglist, (i++ == 0) ? "%s" : ",%s", member->name);
+	}
+	// struct.name=arg1,arg2,argN
+	char *key = r_str_newf ("%s.%s", kind, sname);
+	sdb_set (anal->sdb_types, key, r_strbuf_get (&arglist), 0);
+	free (key);
+
+	key = r_str_newf ("%s.%s.!size", kind, sname);
+	char *val = r_str_newf ("%" PFMT64u "", type->size);
+	sdb_set (anal->sdb_types, key, val, 0);
+	free (val);
+	free (key);
+
+	free (sname);
+
+	r_strbuf_fini (&arglist);
+	r_strbuf_fini (&param_key);
+	r_strbuf_fini (&param_val);
+}
+
+static void save_union(const RAnal *anal, const RAnalBaseType *type) {
+	r_return_if_fail (anal && type && type->name 
+		&& type->kind == R_ANAL_BASE_TYPE_KIND_UNION);
+	const char *kind = "union";
+
+	char *sname = r_str_sanitize_sdb_key (type->name);
+	// name=union
+	sdb_set (anal->sdb_types, sname, kind, 0);
+
+	RStrBuf arglist;
+	RStrBuf param_key;
+	RStrBuf param_val;
+	r_strbuf_init (&arglist);
+	r_strbuf_init (&param_key);
+	r_strbuf_init (&param_val);
+
+	int i = 0;
+	RAnalUnionMember *member;
+	r_vector_foreach (&type->union_data.members, member) {
+		// union.name.arg1=type,offset,???
+		char *member_sname = r_str_sanitize_sdb_key (member->name);
+		r_strbuf_setf (&param_key, "%s.%s.%s", kind, sname, member_sname);
+		r_strbuf_setf (&param_val, "%s,%" PFMT64u ",%" PFMT64u "", member->type, member->offset, 0);
+		sdb_set (anal->sdb_types, r_strbuf_get (&param_key), r_strbuf_get (&param_val), 0);
+		free (member_sname);
+
+		r_strbuf_appendf (&arglist, (i++ == 0) ? "%s" : ",%s", member->name);
+	}
+	// union.name=arg1,arg2,argN
+	char *key = r_str_newf ("%s.%s", kind, sname);
+	sdb_set (anal->sdb_types, key, r_strbuf_get (&arglist), 0);
+	free (key);
+
+	key = r_str_newf ("%s.%s.!size", kind, sname);
+	char *val = r_str_newf ("%" PFMT64u "", type->size);
+	sdb_set (anal->sdb_types, key, val, 0);
+	free (val);
+	free (key);
+
+	free (sname);
+
+	r_strbuf_fini (&arglist);
+	r_strbuf_fini (&param_key);
+	r_strbuf_fini (&param_val);
+}
+
+static void save_enum(const RAnal *anal, const RAnalBaseType *type) {
+	r_return_if_fail (anal && type && type->name 
+		&& type->kind == R_ANAL_BASE_TYPE_KIND_ENUM);
+	/*
+		name=enum
```

Comment:
For these comments, could you please add also the C syntax representation of the type?

Reply:
Sure! Done